rust/processor/src/models/stake_models/current_delegated_voter.rs

```diff
@@ -0,0 +1,351 @@
+#[diesel(table_name = current_delegated_voter)]
+pub struct CurrentDelegatedVoterQuery {
+    pub delegator_address: String,
+    pub delegation_pool_address: String,
+    pub voter: Option<String>,
+    pub pending_voter: Option<String>,
+    pub transaction_timestamp: chrono::NaiveDateTime,
+    pub last_transaction_version: i64,
+    pub table_handle: String,
+}
+
+#[derive(Debug, Deserialize, FieldCount, Identifiable, Insertable, Serialize)]
+#[diesel(primary_key(delegator_address, delegation_pool_address))]
+#[diesel(table_name = current_delegated_voter)]
+pub struct CurrentDelegatedVoter {
+    pub delegator_address: String,
+    pub delegation_pool_address: String,
+    pub voter: Option<String>,
+    pub pending_voter: Option<String>, // voter to be in the next lockup period
+    pub transaction_timestamp: chrono::NaiveDateTime,
+    pub last_transaction_version: i64,
+    pub table_handle: String, // vote_delegation table handle
+}
+
+// (delegation_pool_address, delegator_address)
+pub type CurrentDelegatedVoterPK = (String, String);
+
+pub type CurrentDelegatedVoterMap = HashMap<CurrentDelegatedVoterPK, CurrentDelegatedVoter>;
+// table handle to delegation pool address mapping
+pub type VoteDelegationTableHandleToDelegationPoolAddress = HashMap<String, String>;
+// (optional voter, optional pending voter, table handle)
+pub type PKtoVoteDelegation = HashMap<CurrentDelegatedVoterPK, (Option<String>, Option<String>, String)>;
+
+impl CurrentDelegatedVoter {
+    pub fn get_delegation_pool_address_by_table_handle (
+        conn: &mut PgPoolConnection,
+        table_handle: &str,
+    ) -> anyhow::Result<String> {
+        let mut retried = 0;
+        while retried < QUERY_RETRIES {
+            retried += 1;
+            match CurrentDelegatedVoterQuery::get_by_table_handle (conn, table_handle) {
+                Ok(current_delegated_voter_query_result) => return Ok(current_delegated_voter_query_result.delegation_pool_address),
+                Err(_) => {
+                    std::thread::sleep(std::time::Duration::from_millis(QUERY_RETRY_DELAY_MS));
+                },
+            }
+        }
+        Err(anyhow::anyhow!(
+            "Failed to get delegation pool address from vote delegation write table handle"
+        ))
+    }
+
+    pub fn get_existance_by_pk (
+        conn: &mut PgPoolConnection,
+        delegation_pool_address: &str,
+        delegator_address: &str,
+    ) -> anyhow::Result<bool> {
+        let mut retried = 0;
+        while retried < QUERY_RETRIES {
+            retried += 1;
+            match CurrentDelegatedVoterQuery::get_by_pk (conn, delegation_pool_address, delegator_address) {
+                Ok(current_delegated_voter_query_result) => return Ok(true),
+                Err(_) => {
+                    std::thread::sleep(std::time::Duration::from_millis(QUERY_RETRY_DELAY_MS));
+                },
+            }
+        }
+        Ok(false)
+    }
```

Comment:
Nit: move these at the end of the file? They're not the most important part of this file. I usually prefer having the parsing logic first.